# Mobile App selection data analysis

* What the project is about?
The project is about analysing mobile app data to find out free apps that will help the management decide what type of app to build.

* What your goal is in this project.?
Since we want to build free apps with in-app ads our goal is to decide an app that will be downloaded more. Numbers matter for us.

In [2]:
apple_store = open('AppleStore.csv')
google_store = open('googleplaystore.csv')

from csv import reader

dataset_apple_read = reader(apple_store)
dataset_google_read = reader(google_store)


In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


In [4]:
apple_dataset = list(dataset_apple_read)
appleDataset_header = apple_dataset[0]
apple_dataset = apple_dataset[1:]
explore_data(apple_dataset, 1, 5, True)


print("Coulumns name for apple data" ) 
print(appleDataset_header)
print("\n")


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16
Coulumns name for apple data
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




We have 7197 iOS apps in this data set, and the columns that seem interesting are: 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'. Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home).

In [5]:
google_dataset = list(dataset_google_read)
googleDatatset_header = google_dataset[0]
google_dataset = google_dataset[1:]
explore_data(google_dataset, 1, 5, True)

print("Coulumns name for google data")
print(googleDatatset_header)
print("\n")

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13
Coulumns name for google data
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




We see that the Google Play data set has 10841 apps and 13 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

In [6]:
headerlength = len(googleDatatset_header)
for row in google_dataset:
    rowlength = len(row) 
    if rowlength != headerlength:
        print(row)
        print(google_dataset.index(row))
        del google_dataset[google_dataset.index(row)]

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


In [7]:
print(len(google_dataset))

10840


Google data has a number of duplicate rows. We will print them out for us to see.

In [8]:
G_unique_list = []
G_duplicate_list = []
for row in google_dataset:
    if row[0] in G_unique_list:
        G_duplicate_list.append(row[0])
    else:
        G_unique_list.append(row[0])
print(len(G_unique_list))
print(len(G_duplicate_list))
print("Some of the duplicate rows are:")
print(G_duplicate_list[ :25])

9659
1181
Some of the duplicate rows are:
['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents']


In [9]:
print (googleDatatset_header)
print ("the above is header of the google data set")
for app in google_dataset:
    name = app[0]
    if name == 'Instagram':
        print(app)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
the above is header of the google data set
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Now we need to remove these duplicates, pleaes note that there are more than one duplicates for the same name. We wouldn't remove just anyone of them we would like to keep the one with highest number of reviews and remove the rest.

You can see a sample above (Instagram) it has 4 duplicates but the reviews count is different. We want to keep the one with highest reviews.

The following is the dictionary that will store only unique names with highest reviews

In [10]:
reviews_max = {}
for row in google_dataset:
    name = row[0]
    n_reviews = float(row[3])
    if (name in reviews_max) and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Let's check the lengt of the dictionary and make sure we got the correct number of unique entries that we know.

In [11]:
print("Expected Length:", len(google_dataset) - 1181)
print("Real Length:" ,len(reviews_max))

Expected Length: 9659
Real Length: 9659


let's create 2 list (not dictionary) where the our new cleaned unique dataset will be stored and another list for store app names. 

We will collect the google dataset clean list from the dictionary we crated above. 

In [12]:
google_dataset_clean = [] #(another name android_clean)will store our new cleaned data set 
already_added = [] #will just store app names

for row in google_dataset:
    name = row[0]
    n_reviews = float(row[3])
    #first check dictionary review value matches with the review or not 
    #then see if the name already taken or not of that matched review
    #now if the matched review value's name is already not in already added
    #in the clean list then add the whole row and also add it to that tracker already_added list
    if (reviews_max[name] == n_reviews) and (name not in already_added): 
        google_dataset_clean.append(row)
        already_added.append(name)

In [13]:
print(len(google_dataset_clean))

9659


In [14]:
def english_checker(string):
    for character in string:
        if ord(character) > 127:
            return False
    return True

In [15]:
print (english_checker('Instagram'))
print (english_checker('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print (english_checker('Docs To Go™ Free Office Suite'))
print (english_checker('Instachat 😜'))

True
False
False
False


The program above working fine but it's recognising ™ as non English an emoji as non English. This will lead to losing many English apps. so we improve it below.

In [16]:
def english_checker2(string):
    counter = 0
    for character in string:
        if ord(character) > 127:
            counter+=1
    if counter > 3:
        return False
    else:
        return True

In [17]:
print (english_checker2('Instagram'))
print (english_checker2('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print (english_checker2('Docs To Go™ Free Office Suite'))
print (english_checker2('Instachat 😜'))

True
False
True
True


In [18]:
google_dataset_clean_eng = []

for a_list in google_dataset_clean:
    name = a_list[0]
    if english_checker2(name):
        google_dataset_clean_eng.append(a_list)
print("English Ios Data count",len(google_dataset_clean_eng))

English Ios Data count 9614


In [19]:
apple_dataset_eng = []
for a_list in apple_dataset:
    name = a_list[1]
    if english_checker2(name):
        apple_dataset_eng.append(a_list)
print("English Ios Data count", len(apple_dataset_eng))

English Ios Data count 6183


In [20]:
apple_dataset_eng_free = []
for free_app in apple_dataset_eng:
    price = free_app[4]
    if price == '0.0' :
        apple_dataset_eng_free.append(free_app)
print("Free Ios App count is", len(apple_dataset_eng_free))

Free Ios App count is 3222


In [21]:
google_dataset_clean_eng_free = []
for free_app in google_dataset_clean_eng:
    price = free_app[7]
    if price == '0' :
        google_dataset_clean_eng_free.append(free_app)
print("Free android App count is", len(google_dataset_clean_eng_free))

Free android App count is 8864


We're left with 8864 Android apps and 3222 iOS apps, which should be good enough for our **analysis**.

we want to build an app that will fit in both markets since we want to build android app first. Which will be in minimal version then we want to build a better version and if that version is profitable for 6 months we will build an ios version. 

for that we need to know which genres are common in both markets. We will be building a frequency table for that. 

In [22]:
#analysing the columns and some of the data examples
print("Columns of android", googleDatatset_header)
explore_data(google_dataset_clean_eng_free, 0, 5)
print('\n\n')
print("Columns of ios", appleDataset_header)
explore_data(apple_dataset_eng_free, 0, 5)

Columns of android ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'F

In [23]:
def freq_table(dataset, index):
    freq_dict = {}
    total = 0
    for row in dataset:
        total += 1
        column_value = row[index]
        if column_value in freq_dict:
            freq_dict[column_value] += 1
        else:
            freq_dict[column_value] = 1
    freq_perc = {}
    for key in freq_dict:
        percentage = round((freq_dict[key]/total)*100, 2)
        freq_perc[key] = percentage
    
    return freq_perc
freq_table(apple_dataset_eng_free, 11)


{'Book': 0.43,
 'Business': 0.53,
 'Catalogs': 0.12,
 'Education': 3.66,
 'Entertainment': 7.88,
 'Finance': 1.12,
 'Food & Drink': 0.81,
 'Games': 58.16,
 'Health & Fitness': 2.02,
 'Lifestyle': 1.58,
 'Medical': 0.19,
 'Music': 2.05,
 'Navigation': 0.19,
 'News': 1.33,
 'Photo & Video': 4.97,
 'Productivity': 1.74,
 'Reference': 0.56,
 'Shopping': 2.61,
 'Social Networking': 3.29,
 'Sports': 2.14,
 'Travel': 1.24,
 'Utilities': 2.51,
 'Weather': 0.87}

In [24]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

print("ios Prime Genre")      
display_table(apple_dataset_eng_free, 11)
print("\n\nAndroid Categories")
display_table(google_dataset_clean_eng_free, 1)
print("\n\nAndroid Genres")
display_table(google_dataset_clean_eng_free, 9)

ios Prime Genre
Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


Android Categories
FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AN

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

In [25]:
genres_ios = freq_table(apple_dataset_eng_free, 11)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for genre_app in apple_dataset_eng_free:
        if genre_app[11] == genre:
            total += float(genre_app[5]) #rating_count_tot           
            len_genre += 1
    avg_ratings = total / len_genre
    print(genre, ':', avg_ratings)

Travel : 28243.8
Health & Fitness : 23298.015384615384
Entertainment : 14029.830708661417
Games : 22788.6696905016
Book : 39758.5
Music : 57326.530303030304
Medical : 612.0
Social Networking : 71548.34905660378
Shopping : 26919.690476190477
News : 21248.023255813954
Lifestyle : 16485.764705882353
Utilities : 18684.456790123455
Finance : 31467.944444444445
Photo & Video : 28441.54375
Weather : 52279.892857142855
Navigation : 86090.33333333333
Education : 7003.983050847458
Business : 7491.117647058823
Productivity : 21028.410714285714
Food & Drink : 33333.92307692308
Catalogs : 4004.0
Reference : 74942.11111111111
Sports : 23008.898550724636


Social Networking is dominated by facebook and Navigation i by google maps and waze, Since we are not aiming high at this monent. We could go with the ones that are in 20K-30Ks in avg reviews.

In [26]:
display_table(google_dataset_clean_eng_free, 5) 

1,000,000+ : 15.73
100,000+ : 11.55
10,000,000+ : 10.55
10,000+ : 10.2
1,000+ : 8.39
100+ : 6.92
5,000,000+ : 6.83
500,000+ : 5.56
50,000+ : 4.77
5,000+ : 4.51
10+ : 3.54
500+ : 3.25
50,000,000+ : 2.3
100,000,000+ : 2.13
50+ : 1.92
5+ : 0.79
1+ : 0.51
500,000,000+ : 0.27
1,000,000,000+ : 0.23
0+ : 0.05
0 : 0.01


In [29]:
categories_android = freq_table(google_dataset_clean_eng_free, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in google_dataset_clean_eng_free:
        if app[1] == category:
            n_installs = app[5]
            n_installs = n_installs.replace('+','')
            n_installs = n_installs.replace(',','')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs  )

ART_AND_DESIGN : 1986335.0877192982
COMICS : 817657.2727272727
WEATHER : 5074486.197183099
PARENTING : 542603.6206896552
EVENTS : 253542.22222222222
FAMILY : 3695641.8198090694
BOOKS_AND_REFERENCE : 8767811.894736841
VIDEO_PLAYERS : 24727872.452830188
DATING : 854028.8303030303
FINANCE : 1387692.475609756
HEALTH_AND_FITNESS : 4188821.9853479853
BUSINESS : 1712290.1474201474
EDUCATION : 1833495.145631068
FOOD_AND_DRINK : 1924897.7363636363
MAPS_AND_NAVIGATION : 4056941.7741935486
BEAUTY : 513151.88679245283
SHOPPING : 7036877.311557789
HOUSE_AND_HOME : 1331540.5616438356
PRODUCTIVITY : 16787331.344927534
MEDICAL : 120550.61980830671
GAME : 15588015.603248259
LIFESTYLE : 1437816.2687861272
AUTO_AND_VEHICLES : 647317.8170731707
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
NEWS_AND_MAGAZINES : 9549178.467741935
TOOLS : 10801391.298666667
ENTERTAINMENT : 11640705.88235294
COMMUNICATION : 38456119.167247385
LIBRARIES_AND_DEMO : 638503.734939759
PERSONALIZATION : 5201482.